# Exercise 4.8 Task

## This script will cover the following:
### Step 1 - Create Notebook and Import Data
### Step 2 - Aggregate the Mean of order_id grouped by each Column in a subset

### Step 3 - Analyze the Result
### Step 4 - Create loyalty flag using transform() and loc()
### Step 5 - Spending Habits by Customer Loyalty Type (with troubleshooting for prices column included)
### Step 6 - Spending Flag
### Step 7 - Determining Frequent v. Non-Frequent Users of the App
### Step 8 - Export Data

In [1]:
# Import Libraries

import os
import pandas as pd
import numpy as np

In [2]:
# Set Path

path = r'C:\Users\Josh Wattay\anaconda3\Instacart Basket Analysis'

In [3]:
# Import Data

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl'))

In [4]:
# Check output

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_time_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc    busiest_days  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product    Slowest days   
2         Soda  ...    both  Mid-range product    Slowest days   
3         Soda  ...    both  Mid-range product    Slowest days   
4         Soda  ...    both  Mid-range product    Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_purchase_cost  \
0        Average Orders        10  New customer              6.367535   
1        Average Orders        10  New customer              6.367535   
2           Most Orders        10  New customer              6.367535   
3        Average Orders        10  New customer              6.367535   
4           Most Orders        10  New customer              6.367535   

   spending_flag median_days_btwn_orders   order_frequency_flag  
0    Low spender                    20.5  Non-frequent customer  
1    Low spender                    20.5  Non-frequent customer  
2    Low spender                    20.5  Non-frequent customer  
3    Low spender                    20.5  Non-frequent customer  
4    Low spender                    20.5  Non-frequent customer  

[5 rows x 23 columns]

In [5]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          orders_time_of_day  days_since_prior_order  product_id  \
32404854                  15                     1.0       43553   
32404855                  11                     3.0       43553   
32404856                   7                     NaN       42338   
32404857                   5                    15.0       42338   
32404858                  14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  ...  \
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          _merge    price_range_loc    busiest_days busiest_period_of_day  \
32404854    both  Low-range product  Regularly busy           Most Orders   
32404855    both  Low-range product  Regularly busy           Most Orders   
32404856    both  Mid-range product  Regularly busy        Average Orders   
32404857    both  Mid-range product    Slowest days         Fewest Orders   
32404858    both  Mid-range product    Slowest days           Most Orders   

         max_order      loyalty_flag average_purchase_cost  spending_flag  \
32404854        31  Regular customer              6.905718    Low spender   
32404855        31  Regular customer              6.905718    Low spender   
32404856         3      New customer              7.631527    Low spender   
32404857         3      New customer              7.631527    Low spender   
32404858        16  Regular customer              7.684844    Low spender   

         median_days_btwn_orders  order_frequency_flag  
32404854                     8.0     Frequent customer  
32404855                     8.0     Frequent customer  
32404856                    15.0      Regular customer  
32404857                    15.0      Regular customer  
32404858                    13.0      Regular customer  

[5 rows x 23 columns]

In [6]:
# Shape Check

ords_prods_merge.shape

(32404859, 23)

### Step 2 - Aggregate the Mean of order_number Grouped by department_id for entire dataset

In [7]:
# Apply groupby() and agg() function

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3 - Analyze the Result

#### This result is notably different from the result of a subset created with 1 million records. 
#### Firstly, all 21 departments are represented in this result. 
#### Secondly, the averages for the departments that were represented in the subset are all different.
#### This makes sense, given that we are aggregating about 32.4 times the amount of records as the 1 million record subset. 

### Step 4 - Create loyalty flag using transform() and loc()

In [8]:
# These steps were completed during the exercise using the following code (which I will not repeat hear for practical purposes)

#Creating a column for max_order using transform()

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

#Creating a loyalty Flag using loc()

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_5224\3045331134.py:5: FutureWarning: The provided callable <function amax at 0x0000028CF0ACC220> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [9]:
# Frequency Check

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [10]:
# Aggregate Check using Sum of Results

15876776 + 10284093 + 6243990

32404859

In [11]:
# Shape Check to Verify Aggregate Check Results

ords_prods_merge.shape

(32404859, 23)

The 32,404,859 records from the shape check MATCH the results from our aggregate check. Therefore, all records are accounted for.

### Step 5 - Spending Habits by Customer Loyalty Type (with prices column troubleshooting)

#### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [12]:
# Lets group by "loyalty_flag" and find the mean/average, minimum, and maximum values for the "prices" column

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.775655  1.0  25.0
New customer      7.804260  1.0  25.0
Regular customer  7.801007  1.0  25.0

In [13]:
# Check for division by zero

if (ords_prods_merge['prices'] == 0).any():
    print("Warning: Division by zero may occur. Check your data.")

In [14]:
# Check data type of prices column

ords_prods_merge['prices'].dtype

dtype('float16')

This explains why we were having infinite results when trying to calculate the mean.
I will now change the datatype to int32 and reattempt.

In [15]:
# Verify descriptive stats of prices column

ords_prods_merge['prices'].describe()

C:\Users\Josh Wattay\anaconda3\Lib\site-packages\pandas\core\nanops.py:1479: RuntimeWarning: overflow encountered in cast
  return dtype.type(n)
C:\Users\Josh Wattay\anaconda3\Lib\site-packages\numpy\core\_methods.py:49: RuntimeWarning: overflow encountered in reduce
  return umr_sum(a, axis, dtype, out, keepdims, initial, where)
C:\Users\Josh Wattay\anaconda3\Lib\site-packages\pandas\core\nanops.py:731: RuntimeWarning: invalid value encountered in scalar divide
  the_mean = the_sum / count if count > 0 else np.nan
C:\Users\Josh Wattay\anaconda3\Lib\site-packages\pandas\core\nanops.py:1479: RuntimeWarning: overflow encountered in cast
  return dtype.type(n)


count    3.240486e+07
mean              NaN
std      0.000000e+00
min      1.000000e+00
25%      4.199219e+00
50%      7.398438e+00
75%      1.129688e+01
max      2.500000e+01
Name: prices, dtype: float64

In [16]:
ords_prods_merge['prices'].max()

25.0

In [17]:
mask_inf_values = ords_prods_merge['prices'] == float('inf')

In [18]:
# Creating a boolean mask for rows with infinite values in the prices column

In [19]:
mask_inf_values.value_counts()

prices
False    32404859
Name: count, dtype: int64

In [20]:
# Use the boolean mask to filter the DataFrame

rows_with_inf_values = ords_prods_merge[mask_inf_values]

In [21]:
# Display the rows with infinite values

print(rows_with_inf_values)

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, orders_time_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, product_name, aisle_id, department_id, prices, _merge, price_range_loc, busiest_days, busiest_period_of_day, max_order, loyalty_flag, average_purchase_cost, spending_flag, median_days_btwn_orders, order_frequency_flag]
Index: []

[0 rows x 23 columns]


In [22]:
# Sort the DataFrame by the 'prices' column in descending order

df_sorted = ords_prods_merge.sort_values(by='prices', ascending=False)

In [23]:
df_sorted.head(5130)

order_id  user_id  order_number  orders_day_of_week  \
22675639   2346836   104650            18                   0   
18506618   2551921    34636             4                   2   
18506620    576874    34643             3                   3   
18506621    950934    34643            11                   1   
18506622    613500    34643            12                   2   
...            ...      ...           ...                 ...   
18508245   1657919    65038            13                   0   
18508246    592147    65082             8                   1   
18508247   2671829    65084             5                   2   
18508248   1021170    65084             7                   4   
18508249   3073224    65086             2                   1   

          orders_time_of_day  days_since_prior_order  product_id  \
22675639                  11                    26.0       40486   
18506618                   8                    18.0        9020   
18506620                  20                    24.0        9020   
18506621                   9                     8.0        9020   
18506622                  11                     8.0        9020   
...                      ...                     ...         ...   
18508245                  21                     3.0        9020   
18508246                  13                     7.0        9020   
18508247                  16                     9.0        9020   
18508248                  12                     8.0        9020   
18508249                  16                     8.0        9020   

          add_to_cart_order  reordered                      product_name  ...  \
22675639                 43          0                   Chicken Tenders  ...   
18506618                  9          0  Boneless Skinless Chicken Thighs  ...   
18506620                  3          0  Boneless Skinless Chicken Thighs  ...   
18506621                 22          1  Boneless Skinless Chicken Thighs  ...   
18506622                  9          1  Boneless Skinless Chicken Thighs  ...   
...                     ...        ...                               ...  ...   
18508245                 11          1  Boneless Skinless Chicken Thighs  ...   
18508246                 15          0  Boneless Skinless Chicken Thighs  ...   
18508247                 19          0  Boneless Skinless Chicken Thighs  ...   
18508248                 14          1  Boneless Skinless Chicken Thighs  ...   
18508249                  1          0  Boneless Skinless Chicken Thighs  ...   

          _merge     price_range_loc    busiest_days busiest_period_of_day  \
22675639    both  High-range product    Busiest days           Most Orders   
18506618    both  High-range product  Regularly busy        Average Orders   
18506620    both  High-range product    Slowest days        Average Orders   
18506621    both  High-range product    Busiest days           Most Orders   
18506622    both  High-range product  Regularly busy           Most Orders   
...          ...                 ...             ...                   ...   
18508245    both  High-range product    Busiest days        Average Orders   
18508246    both  High-range product    Busiest days           Most Orders   
18508247    both  High-range product  Regularly busy           Most Orders   
18508248    both  High-range product    Slowest days           Most Orders   
18508249    both  High-range product    Busiest days           Most Orders   

         max_order      loyalty_flag average_purchase_cost  spending_flag  \
22675639        22  Regular customer              7.148770    Low spender   
18506618         7      New customer              7.430911    Low spender   
18506620        27  Regular customer              9.028491    Low spender   
18506621        27  Regular customer              9.028491    Low spender   
18506622        27  Regular customer              9.028491    Low spender   
...            ...               ...    

In [24]:
mask_product_id_21553 = ords_prods_merge['product_id'] == 21553

In [25]:
# Use the boolean mask to filter the DataFrame

records_with_product_id_21553 = ords_prods_merge[mask_product_id_21553]

In [26]:
print(records_with_product_id_21553)

          order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
10034769   3172853   205650            18                   1   
10034770   2504315   205818             3                   5   
10034771   1108388   205818             5                   4   
10034772   1916142   206049             1                   2   
10034773    379732   206049             4                   1   

          orders_time_of_day  days_since_prior_order  product_id  \
10030345                  14                     5.0       21553   
10030346                  16                     4.0       21553   
10030347       

In [27]:
# Update the price values for rows with infinite values

# I will replace 'new_value' with the desired value for these records, wich is the highest non-infinite value of 14896.0

new_value = 14896.0  # Replace with desired value
ords_prods_merge.loc[mask_inf_values, 'prices'] = new_value

In [28]:
# Create a boolean mask for rows with price equal to 14896.0
mask_price_14896 = ords_prods_merge['prices'] == 14896.0

In [29]:
# Update the price values for rows with price equal to 14896.0
new_price_value = 25.00
ords_prods_merge.loc[mask_price_14896, 'prices'] = new_price_value

#### I have discovered a problem in the prices column. 
#### The 2% Milk has an infinite value in the prices column.
#### The Cottage Cheese has a value of 14896. Obviously this is an error, or the worlds best cottage cheese.
#### Therefore, above I used a boolean mask technique to locate and change the infinite values of the Milk to the values of the Cottage Cheese.
#### I then used the same technique to change the erroneous value of the milk and cottage cheese from 14896 to 25.
#### The value of 25 was determined by searching through the dataframe for the next highest value below 14896 by using the head() function to scan.
#### I will now re-attempt finding aggregate values for prices grouped by customer loyalty type.

In [30]:
# Re-attempt to find mean, min, and max for customer types 

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.775655  1.0  25.0
New customer      7.804260  1.0  25.0
Regular customer  7.801007  1.0  25.0

#### Based on our data, there is no noticeable difference between the spending habits of customers based on the loyalty_flag criteria.

### Step 6 - Spending Flag

#### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

#### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

#### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [31]:
# Lets start by using the transform() function to determine the mean price of products based on each user

# We will create a new column with this information called 'average_purchase_cost'

ords_prods_merge['average_purchase_cost'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_5224\2522240418.py:5: FutureWarning: The provided callable <function mean at 0x0000028CF0ACCF40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_purchase_cost'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [32]:
# Check output for user_id and new column

ords_prods_merge[['user_id', 'average_purchase_cost']].head(50)

user_id  average_purchase_cost
0         1               6.367535
1         1               6.367535
2         1               6.367535
3         1               6.367535
4         1               6.367535
5         1               6.367535
6         1               6.367535
7         1               6.367535
8         1               6.367535
9         1               6.367535
10       15               3.980306
11       15               3.980306
12       15               3.980306
13       15               3.980306
14       15               3.980306
15       19               8.372018
16       19               8.372018
17       19               8.372018
18       21               7.012538
19       31               8.137767
20       31               8.137767
21       43               7.444980
22       43               7.444980
23       52               6.356293
24       52               6.356293
25       52               6.356293
26       52               6.356293
27       52               6.356293
28       52               6.356293
29       52               6.356293
30       52               6.356293
31       52               6.356293
32       52               6.356293
33       52               6.356293
34       52               6.356293
35       52               6.356293
36       52               6.356293
37       67               7.656539
38       67               7.656539
39       67               7.656539
40       67               7.656539
41       67               7.656539
42       67               7.656539
43       67               7.656539
44       67               7.656539
45       67               7.656539
46       67               7.656539
47       67               7.656539
48       67               7.656539
49       67               7.656539

#### Now we can use the loc() function to create a Spending Flag that meets the criteria provided by the stakeholders.

In [33]:
# For High spender category

ords_prods_merge.loc[ords_prods_merge['average_purchase_cost'] >= 10, 'spending_flag'] = 'High spender'

In [34]:
# For Low spender category

ords_prods_merge.loc[ords_prods_merge['average_purchase_cost'] < 10, 'spending_flag'] = 'Low spender'

In [35]:
# Check output of columns

ords_prods_merge[['user_id', 'average_purchase_cost', 'spending_flag']].head(50)

user_id  average_purchase_cost spending_flag
0         1               6.367535   Low spender
1         1               6.367535   Low spender
2         1               6.367535   Low spender
3         1               6.367535   Low spender
4         1               6.367535   Low spender
5         1               6.367535   Low spender
6         1               6.367535   Low spender
7         1               6.367535   Low spender
8         1               6.367535   Low spender
9         1               6.367535   Low spender
10       15               3.980306   Low spender
11       15               3.980306   Low spender
12       15               3.980306   Low spender
13       15               3.980306   Low spender
14       15               3.980306   Low spender
15       19               8.372018   Low spender
16       19               8.372018   Low spender
17       19               8.372018   Low spender
18       21               7.012538   Low spender
19       31               8.137767   Low spender
20       31               8.137767   Low spender
21       43               7.444980   Low spender
22       43               7.444980   Low spender
23       52               6.356293   Low spender
24       52               6.356293   Low spender
25       52               6.356293   Low spender
26       52               6.356293   Low spender
27       52               6.356293   Low spender
28       52               6.356293   Low spender
29       52               6.356293   Low spender
30       52               6.356293   Low spender
31       52               6.356293   Low spender
32       52               6.356293   Low spender
33       52               6.356293   Low spender
34       52               6.356293   Low spender
35       52               6.356293   Low spender
36       52               6.356293   Low spender
37       67               7.656539   Low spender
38       67               7.656539   Low spender
39       67               7.656539   Low spender
40       67               7.656539   Low spender
41       67               7.656539   Low spender
42       67               7.656539   Low spender
43       67               7.656539   Low spender
44       67               7.656539   Low spender
45       67               7.656539   Low spender
46       67               7.656539   Low spender
47       67               7.656539   Low spender
48       67               7.656539   Low spender
49       67               7.656539   Low spender

In [36]:
# Frequency Check

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32282827
High spender      122032
Name: count, dtype: int64

In [37]:
# Aggregate Check
32282827 + 122032

32404859

In [38]:
# Shape Check to verify Aggregate Check

ords_prods_merge.shape

(32404859, 23)

The 32,404,859 result of the Aggregate check MATCHES our dataframe Shape check. Therefore, all records are accounted for.

A vast majority of users average purchase is less than $10.

In [39]:
ords_prods_spending_flag = ords_prods_merge['spending_flag'].value_counts()

ords_prods_spending_flag.to_clipboard()

### Step 7 - Determining Frequent v. Non-Frequent Users of the App

#### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

#### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a "Regular customer."
#### If the median is lower than or equal to 10, then the customer should be labeled a "Frequent customer.".”

In [40]:
# Lets start by using transform() again with a groupby() to determine a users median days since prior order

# We will create a median_days_btwn_orders column to give each user the median values needed to qualify for the Order Frequency Flag

ords_prods_merge['median_days_btwn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_5224\3475829972.py:5: FutureWarning: The provided callable <function median at 0x0000028CF0C1B6A0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_btwn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [41]:
# Check output for user_id and new column

ords_prods_merge[['user_id', 'median_days_btwn_orders']].head(50)

user_id  median_days_btwn_orders
0         1                     20.5
1         1                     20.5
2         1                     20.5
3         1                     20.5
4         1                     20.5
5         1                     20.5
6         1                     20.5
7         1                     20.5
8         1                     20.5
9         1                     20.5
10       15                     10.0
11       15                     10.0
12       15                     10.0
13       15                     10.0
14       15                     10.0
15       19                      6.0
16       19                      6.0
17       19                      6.0
18       21                      7.5
19       31                      6.0
20       31                      6.0
21       43                     10.0
22       43                     10.0
23       52                      8.0
24       52                      8.0
25       52                      8.0
26       52                      8.0
27       52                      8.0
28       52                      8.0
29       52                      8.0
30       52                      8.0
31       52                      8.0
32       52                      8.0
33       52                      8.0
34       52                      8.0
35       52                      8.0
36       52                      8.0
37       67                      7.0
38       67                      7.0
39       67                      7.0
40       67                      7.0
41       67                      7.0
42       67                      7.0
43       67                      7.0
44       67                      7.0
45       67                      7.0
46       67                      7.0
47       67                      7.0
48       67                      7.0
49       67                      7.0

#### Now we can use the loc() function to set the conditions for our Order Frequency Flag.

In [42]:
# For Non-frequent customer

ords_prods_merge.loc[ords_prods_merge['median_days_btwn_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [43]:
# For Regular customer

ords_prods_merge.loc[(ords_prods_merge['median_days_btwn_orders'] > 10) & (ords_prods_merge['median_days_btwn_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [44]:
# For Frequent customer

ords_prods_merge.loc[ords_prods_merge['median_days_btwn_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [45]:
# Check output for new order_frequency_flag column with user_id and median_days_btwn_orders

ords_prods_merge[['user_id', 'median_days_btwn_orders', 'order_frequency_flag']].head(50)

user_id  median_days_btwn_orders   order_frequency_flag
0         1                     20.5  Non-frequent customer
1         1                     20.5  Non-frequent customer
2         1                     20.5  Non-frequent customer
3         1                     20.5  Non-frequent customer
4         1                     20.5  Non-frequent customer
5         1                     20.5  Non-frequent customer
6         1                     20.5  Non-frequent customer
7         1                     20.5  Non-frequent customer
8         1                     20.5  Non-frequent customer
9         1                     20.5  Non-frequent customer
10       15                     10.0      Frequent customer
11       15                     10.0      Frequent customer
12       15                     10.0      Frequent customer
13       15                     10.0      Frequent customer
14       15                     10.0      Frequent customer
15       19                      6.0      Frequent customer
16       19                      6.0      Frequent customer
17       19                      6.0      Frequent customer
18       21                      7.5      Frequent customer
19       31                      6.0      Frequent customer
20       31                      6.0      Frequent customer
21       43                     10.0      Frequent customer
22       43                     10.0      Frequent customer
23       52                      8.0      Frequent customer
24       52                      8.0      Frequent customer
25       52                      8.0      Frequent customer
26       52                      8.0      Frequent customer
27       52                      8.0      Frequent customer
28       52                      8.0      Frequent customer
29       52                      8.0      Frequent customer
30       52                      8.0      Frequent customer
31       52                      8.0      Frequent customer
32       52                      8.0      Frequent customer
33       52                      8.0      Frequent customer
34       52                      8.0      Frequent customer
35       52                      8.0      Frequent customer
36       52                      8.0      Frequent customer
37       67                      7.0      Frequent customer
38       67                      7.0      Frequent customer
39       67                      7.0      Frequent customer
40       67                      7.0      Frequent customer
41       67                      7.0      Frequent customer
42       67                      7.0      Frequent customer
43       67                      7.0      Frequent customer
44       67                      7.0      Frequent customer
45       67                      7.0      Frequent customer
46       67                      7.0      Frequent customer
47       67                      7.0      Frequent customer
48       67                      7.0      Frequent customer
49       67                      7.0      Frequent customer

In [46]:
# Frequency Check

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [49]:
ords_prods_freq_flag = ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

ords_prods_freq_flag.to_clipboard()

In [47]:
# Aggregate Check

21559853 + 7208564 + 3636437 + 5

32404859

In [48]:
# Shape Check to verify Aggregate Check

ords_prods_merge.shape

(32404859, 23)

#### The 32,404,859 rows from our shape check MATCH the results of the aggregate check. Therefore, all records are accounted for.
#### It should be noted that there is a count of 5 in the nan column. This indicates first time purchases by new users who have never made another purchase. 

### 8. Export Data

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4.8.pkl'))